internal/graph/checkingresourcestream.go

```diff
@@ -0,0 +1,537 @@
+		reachableResourcesCompleted: make(chan struct{}, concurrencyLimit),
+		availableForPublishing:      make(chan bool, concurrencyLimit),
+
+		orderingIndexToBePublished: 0,
+		reachableResourcesCount:    0,
+
+		errSetter: sync.Once{},
+		err:       nil,
+
+		processingWaitGroup: sync.WaitGroup{},
+		publishingWaitGroup: sync.WaitGroup{},
+	}
+
+	// Spawn the goroutine that will publish resources to the parent stream in the proper order.
+	crs.publishingWaitGroup.Add(1)
+	go crs.resourcePublisher()
+	return crs
+}
+
+// waitForPublishing waits for the publishing goroutine to complete its work, and returns the number
+// of published *reachable* resources or the error that occurred during checking or publishing.
+func (crs *checkingResourceStream) waitForPublishing() (uint64, *v1.Cursor, error) {
+	// Mark that no new items will come in from the reachable resources stream.
+	for i := 0; i < int(crs.concurrencyLimit); i++ {
+		crs.reachableResourcesCompleted <- struct{}{}
+	}
+
+	// Wait for all processing to complete.
+	crs.processingWaitGroup.Wait()
+
+	// Mark publishing as ready for final publishing.
+	crs.availableForPublishing <- false
+
+	// Wait for any remaining publishing to complete.
+	crs.publishingWaitGroup.Wait()
+
+	return crs.reachableResourcesCount, crs.lastResourceCursor, crs.err
+}
+
+// resourcePublisher is the goroutine that publishes resources to the parent stream once they've been
+// validated by the processing worker(s).
+func (crs *checkingResourceStream) resourcePublisher() {
+	defer crs.publishingWaitGroup.Done()
+
+	for {
+		select {
+		case <-crs.ctx.Done():
+			return
+
+		case isStillRunning := <-crs.availableForPublishing:
+			err := crs.publishResourcesIfPossible()
+			if err != nil {
+				crs.setError(err)
+				return
+			}
+
+			if !isStillRunning {
+				return
+			}
+
+			continue
+		}
+	}
+}
+
+// publishResourcesIfPossible publishes the resources that have been processed, in the correct order, if any.
+func (crs *checkingResourceStream) publishResourcesIfPossible() error {
+	for {
+		toPublish := crs.rq.resourcesToPossiblyPublish()
+		if len(toPublish) == 0 {
+			return nil
+		}
+
+		for {
+			if len(toPublish) == 0 {
+				break
+			}
+
+			// Sort to ensure they are in the publishable order.
+			sort.Sort(byOrderingIndex(toPublish))
+
+			// Ensure that the next resource to be published is the next in the order. If not,
+			// we're still waiting on a resource to be checked.
+			current := toPublish[0]
+			if current.orderingIndex != crs.orderingIndexToBePublished {
+				return nil
+			}
+
+			toPublish = toPublish[1:]
+			crs.orderingIndexToBePublished++
+
+			// NOTE: lookupResult will be `nil` if the Check for the resource found that the resource is
+			// not actually accessible. The entry is kept in `toPublish` to ensure proper ordering is maintained
+			// on the parent stream.
+			if current.lookupResult != nil {
+				ok, done := crs.limits.prepareForPublishing()
+				defer done()
```

Comment:
> but it would have incremented orderingIndexToBePublished and not published the resource.

That's fine: once the limit is reached, no additional publishing should occur.